src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,496 @@
+                    {
+                        return new AppendPrependNIterator<TSource>(_source, _prepended, _preCount + 1, _appended, _appCount);
+                    }
+                }
+
+                return new AppendPrepend1Iterator<TSource>(this, element, false);
+            }
+
+            public override int GetCount(bool onlyIfCheap)
+            {
+                IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                if (listProv != null)
+                {
+                    int count = listProv.GetCount(onlyIfCheap);
+                    return count == -1 ? -1 : count + _preCount + _appCount;
+                }
+
+                return !onlyIfCheap || _source is ICollection<TSource> || _source is ICollection ? _source.Count() + _preCount + _appCount : -1;
+            }
+
+            public override TSource[] ToArray()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count];
+                _prepended?.CopyToFromEnd(array, _preCount);
+                int index = _preCount;
+                ICollection<TSource> sourceCollection = _source as ICollection<TSource>;
+                if (sourceCollection != null)
+                {
+                    sourceCollection.CopyTo(array, index);
+                    index += sourceCollection.Count;
+                }
+                else
+                {
+                    foreach (TSource item in _source)
+                    {
+                        array[index] = item;
+                        ++index;
+                    }
+                }
+
+                _appended?.CopyTo(array, index, _appCount);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(Math.Max(_appCount + _preCount, 4)) : new List<TSource>(count);
+                _prepended?.AddTo(list, _preCount);
+                list.AddRange(_source);
+                _appended?.AddToFromEnd(list, _appCount);
+                return list;
+            }
+        }
+    }
+
+    internal sealed class SharedCollection<T>
+    {
+        // Limit the number of elements stored, because the collection being shared makes it possible
+        // for a series of operations to grow the collection large, but for only one of the earlier
+        // operations to be keeping it alive.
+        private const int MaxLength = 64;
+        private T[] _store;
+        private int _count;
+
+        public SharedCollection(T item)
+        {
+            _store = new[] { item };
+            _count = 1;
+        }
+
+        public SharedCollection(T first, T second)
```

Comment:
Maybe it would be more readable if you had a `params T[] items` constructor here instead?